### PR TITLE
Move Zombie detection to SchedulerJob

### PR DESCRIPTION
**Pull request by mhenc**
Moves zombie detection method from DagProcessor to SchedulerJob.

More context:
https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-43+DAG+Processor+separation

### CODE REVIEW

**Comment by mhenc** [comment]
cc: @potiuk

**Review by potiuk — APPROVED** [review]
This one looks good (and a good start for AIP-43). 

One to add is that we are changing a default value for detection interval from 10 to 30 and there are two questions: 

* can it have some unexpected consequences (I think only positive - i.e. less load, 30s is pretty fast for zombie detection IMHO)
* if we agree to change it to 30 s. it should be added to UPDATING.md as a "change" for the upcoming 2.3.0

Since this is change in the core/scheduler I will need second commiter review here. @ashb @uranusjr @ephraimbuddy  I think you are the "closest" to scheduler_job and you can think if there are some onforeseen consequences. I thin

File: airflow/models/dagbag.py [hunk 1]
```diff
@@ -89,7 +89,6 @@ class DagBag(LoggingMixin):
     """
 
     DAGBAG_IMPORT_TIMEOUT = conf.getfloat('core', 'DAGBAG_IMPORT_TIMEOUT')
-    SCHEDULER_ZOMBIE_TASK_THRESHOLD = conf.getint('scheduler', 'scheduler_zombie_task_threshold')
```
Comment (uranusjr) [inline airflow/models/dagbag.py:92]: Hmm, I wonder if we can remove this. It’s not used anywhere (even before this PR) and is probably kept for backward compatibility?

**Comment by uranusjr** [comment]
Logic makes sense to me although I don’t have much real-world experience to comment on the configuration issue.

File: airflow/models/dagbag.py (hunk 1 above)
Comment (mhenc) [reply airflow/models/dagbag.py:92]: Right, I didn't find any reference to that in the codebase, so I just removed as unused code.

But if you believe it should be left there, then of course I can revert this change

Comment (uranusjr) [reply airflow/models/dagbag.py:92]: Let’s revert to be safe (and add a comment saying this is not actually used anywhere and can be removed in Airflow 3).

**Comment by mhenc** [comment]
If you believe it would be safer to keep 10s then we may use it. Although it sounds little to often for me.
Nonetheless it's now a configuration option, so users are able to set it to the value they like.

**Comment by potiuk** [comment]
> If you believe it would be safer to keep 10s then we may use it. Although it sounds little to often for me. Nonetheless it's now a configuration option, so users are able to set it to the value they like.

I think it's ok to change (I see no serious effects of such change) but note in UPDATING.md will be needed. @ashb  - was there a rationale behind the 10s initially ? Or was it just arbitrary chosen without much of a "reasoning" ?

File: airflow/models/dagbag.py (hunk 1 above)
Comment (potiuk) [reply airflow/models/dagbag.py:92]: I would be for removing it. There is one other place (local_task_job) where the threshold is used (but directly) so I think this is a leftover from separating this on out. I do not imagine any use of the constant by the users :). Anyone who wants this value should do `conf.getint`.

**Comment by ashb** [comment]
No idea the source of the 10s -- it was from 2018 in #3873

**Comment by potiuk** [comment]
> No idea the source of the 10s -- it was from 2018 in #3873

Ah  good one. I should have checked before! @KevinYang21  - maybe you remember / have some insights if the 10s were chosen for a good reason?

**Comment by ashb** [comment]
If you aren't familiar with "git log pickaxe" it's amazing for things like this http://www.philandstuff.com/2014/02/09/git-pickaxe.html

```
airflow ❯ git log -S _zombie_query_interval --oneline 
897960736 Revert "[AIRFLOW-4797] Improve performance and behaviour of zombie detection (#5511)"
2bdb053db [AIRFLOW-4797] Improve performance and behaviour of zombie detection (#5511)
75e2288a3 [Airflow-2760] Decouple DAG parsing loop from scheduler loop (#3873)
```

**Comment by potiuk** [comment]
TIL! thanks @ashb !

File: tests/jobs/test_scheduler_job.py [hunk 2]
```diff
@@ -75,8 +76,7 @@
 )
 PERF_DAGS_FOLDER = os.path.join(ROOT_FOLDER, "tests", "test_utils", "perf", "dags")
 ELASTIC_DAG_FILE = os.path.join(PERF_DAGS_FOLDER, "elastic_dag.py")
-
-TEST_DAG_FOLDER = os.environ['AIRFLOW__CORE__DAGS_FOLDER']
+TEST_DAG_FOLDER = '/airflow/github/dags'
```
Comment (ephraimbuddy) [inline tests/jobs/test_scheduler_job.py]: Why is this change?

Comment (mhenc) [reply tests/jobs/test_scheduler_job.py]: Good catch. It's just some leftover from my tests.
Reverted.

**Comment by mhenc** [comment]
I reverted the interval time to 10 seconds to make it fully backward compatible.
As it's configuration option now, then user may decide to bump up it to save some CPU/DB resources.

File: airflow/jobs/scheduler_job.py [hunk 3]
```diff
@@ -1259,3 +1267,39 @@ def check_trigger_timeouts(self, session: Session = None):
...
         if num_timed_out_tasks:
             self.log.info("Timed out %i deferred tasks without fired triggers", num_timed_out_tasks)
+
+    @provide_session
+    def _find_zombies(self, session):
+        """
+        Find zombie task instances, which are tasks haven't heartbeated for too long
+        and update the current zombie list.
+        """
+        self.log.info("Finding 'running' jobs without a recent heartbeat")
```
Comment (ephraimbuddy) [inline airflow/jobs/scheduler_job.py]: Can we change this to debug or remove it entirely because seeing this every 10 seconds seems much

**Review by ephraimbuddy — APPROVED** [review]

File: airflow/jobs/scheduler_job.py (hunk 3 above)
Comment (mhenc) [reply airflow/jobs/scheduler_job.py]: Sure, changed to debug.

**Review by potiuk — APPROVED** [review]

**Pull request merged**